pymc_experimental/statespace/utils/numba_linalg.py

```diff
@@ -0,0 +1,17 @@
+import numpy as np
+from numba import njit
+
+
+@njit
+def numba_block_diagonal(nd_array):
+    # TODO: Replace this with an overload for scipy.linalg.block_diagonal
+    n, rows, cols = nd_array.shape
+
+    out = np.zeros((n * rows, n * cols))
+
+    r, c = 0, 0
+    for i, (rr, cc) in enumerate([(rows, cols)] * n):
+        out[r : r + rr, c : c + cc] = nd_array[i]
+        r += rr
+        c += cc
+    return out
```

Comment:
Make sure `numba` is still an optional dependency for users of `pymc_experimental`

Reply:
Yeah hopefully this can all just go away by making the filter a distribution.

Do you have any idea how to handle the fact that there are several types of prior/posterior you can sample out of the model: conditional vs unconditional, then the different filter outputs (1-step ahead predictions, filtered, and smoothed), then mean vs covariance? I was playing around with it a bit and it seemed like the random method can't return a tuple -- is that incorrect?

Reply:
It has to return a numpy array. Are you trying to do a multi-output Op?

Reply:
The graph for the kalman filter returns 5 values: a sequence of mean vectors and covariance matrices taking into account information up to time t ("kalman predictions"), and the same sequence also incorporating information from time t itself ("kalman filter"), and the log-likelihood of the data at time t, given the kalman predictions. I'll call these mu_p_t, Sigma_p_t, mu_f_t, Sigma_f_t, and ll_p_t.

So there are two sequences of distributions returned: MVN(mu_p_t, Sigma_p_t), and MVN(mu_f_t, Sigma_f_t). This block_diagonal function arises because I use it to sample an entire trajectory all in one go: MVN(concat(mu_p_t), block_diag(Sigma_p_t)), for example. In principle pymc could also compute ll_p_t this way, but it's a bit wasteful because it requires inversion of a bigger covariance matrix. 

But for the random function, either or both of these stacked MVN distributions could be returned, depending on the situation. Or an "unconditional" distribution could be returned, that just iterates the transition and observation equations without incorporating any observations at all.

For the logp function, on the other hand, only the predicted distribution makes sense. My current thought is to split into 4 distributions: KalmanUnconditionalDist, KalmanFilterDist, KalmanPredictionDist, and KalmanSmootherDist, and use a wrapper that returns all 4. Only the PredicitonDist would have an implemented logp function. Thoughts?

Also, is it necessary to re-write the filter in numpy for the random function? Or can I just use the pytensor code I already have?

Reply:
> For the logp function, on the other hand, only the predicted distribution makes sense. My current thought is to split into 4 distributions: KalmanUnconditionalDist, KalmanFilterDist, KalmanPredictionDist, and KalmanSmootherDist, and use a wrapper that returns all 4. Only the PredicitonDist would have an implemented logp function. Thoughts?

Sounds good

> Also, is it necessary to re-write the filter in numpy for the random function? Or can I just use the pytensor code I already have?

Should be fine to use PyTensor code, but then you wouldn't create a `RandomVariable`. That's the exact case for `SymbolicDisttribution`s.